src/ray/gcs/gcs_client/gcs_client.cc

```diff
@@ -180,10 +180,38 @@ std::pair<std::string, int> GcsClient::GetGcsServerAddress() {
   return current_gcs_server_address_;
 }
 
+/// Checks whether GCS at the specified address is healthy.
+bool GcsClient::CheckHealth(const std::string &ip, int port, int64_t timeout_ms) {
+  // Health checking currently needs to be blocking. So it cannot use the event loop
+  // in the GcsClient. Otherwise there can be deadlocks.
+  auto channel = grpc::CreateChannel(absl::StrCat(ip, ":", port),
+                                     grpc::InsecureChannelCredentials());
+  std::unique_ptr<rpc::HeartbeatInfoGcsService::Stub> stub =
+      rpc::HeartbeatInfoGcsService::NewStub(std::move(channel));
+  grpc::ClientContext context;
+  context.set_deadline(std::chrono::system_clock::now() +
+                       std::chrono::milliseconds(timeout_ms));
+  const rpc::CheckAliveRequest request;
+  rpc::CheckAliveReply reply;
+  auto status = stub->CheckAlive(&context, request, &reply);
+  if (!status.ok()) {
+    RAY_LOG(WARNING) << "Unable to reach GCS at " << ip << ":" << port
+                     << ". Failure: " << status.error_code() << " "
+                     << status.error_message();
+    return false;
+  }
+  return true;
+}
+
 void GcsClient::PeriodicallyCheckGcsServerAddress() {
   if (disconnected_) {
     return;
   }
+  if (current_connection_failure_.has_value()) {
```

Comment:
I feel this logic is a little bit hard to follow. This function seems to mix handling server addr change and connection error and this function is called periodically which doesn't seem very fit doing handling error change.

Maybe we should check the logic here and maybe come up with a better flow of doing the work.

Reply:
Renamed the function to `PeriodicallyCheckGcsConnection()`. The problem with the existing approach to handle GCS client rpc error is that if there is a connection issue, all inflight requests would fail and sequentially each response callback would try to reconnect with GCS. This would slow down the client and may trigger unexpected behavior with the socket connection test. It seems better to have only one callback responsible to trying to reconnect with GCS. Alternatively, we can create another periodic runner for this purpose. Or we can revert to the existing behavior.